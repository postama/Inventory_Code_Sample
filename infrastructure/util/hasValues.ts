export function hasValues(object: object): boolean {
    return object && Object.keys(object).length > 0;
}

export function isEmpty(object: object): boolean {
    return !hasValues(object);
}