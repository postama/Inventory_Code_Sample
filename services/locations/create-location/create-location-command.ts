import { literal, intersection, type, TypeOf } from 'io-ts';
import { BaseLocationCommand } from '../shared/base-location-command';
import { validateType } from '../../../infrastructure/validator/validator';
import uuid from 'uuid/v4';

export const CreateLocationCommandValidator = intersection([
    BaseLocationCommand,
    type({ type: literal('CreateLocation') })
]);

export type CreateLocationCommand = TypeOf<typeof CreateLocationCommandValidator>;

export function createCreateLocationCommand(request: any): CreateLocationCommand {
    return validateType({ ...request, isActive: true, type: 'CreateLocation', locationId: uuid() }, CreateLocationCommandValidator);
}