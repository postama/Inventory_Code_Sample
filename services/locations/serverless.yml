service: inventory-locations

provider:
  name: aws
  runtime: nodejs8.10
  profile: kanso
  stage: local
  region: us-west-2
  environment:
    NODE_CONFIG_ENV: ${self:custom.stage}
  iamRoleStatements:
    - Effect: Allow
      Action:
        - dynamodb:Query
        - dynamodb:Scan
        - dynamodb:GetItem
        - dynamodb:PutItem
      Resource:
        - 'Fn::ImportValue': ${self:custom.stage}-InventoryTableArn
    - Effect: Allow
      Action:
        - dynamodb:Query
        - dynamodb:Scan
      Resource: {"Fn::Join": ["", [{"Fn::ImportValue":"${self:custom.stage}-InventoryTableArn"}, "/index/*"]]}

package:
  individually: ${self:custom.packageIndividually.${self:custom.stage}, self:custom.packageIndividually.other}

plugins:
  - serverless-add-api-key
  - serverless-webpack
  - serverless-offline
  - serverless-domain-manager

custom:
  apiKeys:
    - name: inventory-key
      value: ${ssm:/doorways/${self:custom.stage}/INVENTORY_API_KEY}
  customDomain:
    domainName: ${self:custom.domainPrefix.${self:custom.stage}, self:custom.domainPrefix.other}.doorways-services.net
    certificateName: '*.doorways-services.net'
    basePath: location
  domainPrefix:
    prod: inventory
    other: ${self:custom.stage}-inventory
  stage: ${opt:stage, self:provider.stage}
  region: ${opt:region, self:provider.region}
  packageIndividually:
    local: false
    other: true

functions:
  CreateLocation:
    handler: create-location/create-location.entry
    events:
      - http:
          path: v1
          method: POST
          private: true
  EditLocation:
    handler: edit-location/edit-location.entry
    events:
      - http:
          path: v1/{locationId}
          method: PUT
          private: true