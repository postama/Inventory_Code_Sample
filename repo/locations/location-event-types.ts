import { LocationCreatedV1 } from '../../services/locations/create-location/create-location-event';
import { LocationEditedV1 } from '../../services/locations/edit-location/edit-location-event';

export type LocationEvent =
    LocationCreatedV1 |
    LocationEditedV1;