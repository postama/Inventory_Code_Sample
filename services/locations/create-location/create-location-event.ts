import { intersection, type, literal, TypeOf } from 'io-ts';
import { createLocation, LocationValidator } from '../../../repo/locations/location-type';
import { CreateLocationCommand } from './create-location-command';
import { validateType } from '../../../infrastructure/validator/validator';
import { createBaseEvent, BaseEventValidator } from '../../../repo/events/event-type';
import uuid from 'uuid/v4';

export const CREATE_LOCATION = 'LocationCreated';

export const LocationCreatedValidatorV1 = intersection([
    BaseEventValidator,
    type({
        type: literal(CREATE_LOCATION),
        version: literal(1),
        data: LocationValidator
    })
]);

export type LocationCreatedV1 = TypeOf<typeof LocationCreatedValidatorV1>;

export function createCreateLocationEvent(command: CreateLocationCommand): LocationCreatedV1 {
    const event = {
        ...createBaseEvent(command, uuid(), CREATE_LOCATION, 1),
        data: createLocation(command)
    };

    return validateType(event, LocationCreatedValidatorV1);
}