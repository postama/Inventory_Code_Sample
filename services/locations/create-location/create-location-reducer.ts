import { Location } from '../../../repo/locations/location-type';
import { LocationCreatedV1 } from './create-location-event';

// tslint:disable-next-line variable-name
export function createLocationReducer(_location: Location, event: LocationCreatedV1): Location {
    return {
        ...event.data,
        siteId: event.siteId,
        created: event.created
    };
}