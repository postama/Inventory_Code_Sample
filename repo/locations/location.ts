import * as dynamo from '../../infrastructure/persistence/dynamo';
import { Reducer } from '../../infrastructure/reducer/reducer';
import { Location } from './location-type';
import { LocationEvent } from './location-event-types';
import { CREATE_LOCATION } from '../../services/locations/create-location/create-location-event';
import { createLocationReducer } from '../../services/locations/create-location/create-location-reducer';
import { getEntityQuery } from '../events/entity';
import { EDIT_LOCATION } from '../../services/locations/edit-location/edit-location-event';
import { editLocationReducer } from '../../services/locations/edit-location/edit-location-reducer';

export const LocationReducer = new Reducer<Location, LocationEvent>();

LocationReducer.register(CREATE_LOCATION, createLocationReducer);
LocationReducer.register(EDIT_LOCATION, editLocationReducer);

export async function getById(customerId: string, locationId: string): Promise<Location> {
    return await dynamo.pagedReducer<Location, LocationEvent>(LocationReducer, getEntityQuery(customerId, locationId));
}

export async function save(event: LocationEvent) {
    return await dynamo.put(event);
}