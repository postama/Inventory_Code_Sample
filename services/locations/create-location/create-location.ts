import { APIGatewayEvent } from 'aws-lambda';
import { requestHandler } from '../../../../Kanso.Doorways.InventoryShared/entry-wrapper/entry-wrapper';
import { CreateLocationCommand, createCreateLocationCommand } from './create-location-command';
import { checkUniqueCommand } from '../../../repo/events/event';
import { LocationCreatedV1, createCreateLocationEvent } from './create-location-event';
import { createLocationReducer } from './create-location-reducer';
import { save } from '../../../repo/locations/location';

export function entry(request: APIGatewayEvent) {
    return requestHandler(createLocation, request);
}

export async function createLocation(request): Promise<void> {
    const command: CreateLocationCommand = createCreateLocationCommand(request);
    await checkUniqueCommand(command);
    const event: LocationCreatedV1 = createCreateLocationEvent(command);
    event.state = createLocationReducer({} as any, event);
    await save(event);
    return;
}