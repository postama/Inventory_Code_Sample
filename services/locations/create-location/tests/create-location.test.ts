import { createTestCreateLocationCommand, createTestCreateLocationEvent } from './test.util';
import { createCreateLocationCommand } from '../create-location-command';
import { createCreateLocationEvent } from '../create-location-event';
import { createLocationReducer } from '../create-location-reducer';
import { entry, createLocation } from '../create-location';

import * as entryLib from '../../../../../Kanso.Doorways.InventoryShared/entry-wrapper/entry-wrapper';
jest.mock('../../../../../Kanso.Doorways.InventoryShared/entry-wrapper/entry-wrapper');
const entryMock = entryLib as jest.Mocked<typeof entryLib>;

import * as locationRepo from '../../../../repo/locations/location';
const spy = jest.spyOn(locationRepo, 'save');

describe('create location', () => {
    describe('command structure validator', () => {
        test('should fail without user with a human readable error message', () => {
            const command = createTestCreateLocationCommand({ user: undefined });
            expect.assertions(2);
            try {
                createCreateLocationCommand(command);
            } catch (e) {
                expect(e.name).toBe('ValidationError');
                expect(e.message).toBe('Expecting string at user but instead got: undefined.');
            }
        });

        test('should pass with full structure', () => {
            const command = createTestCreateLocationCommand();
            const validCommand = createCreateLocationCommand(command);
            expect(validCommand).toMatchSnapshot({
                locationId: expect.any(String),
                customerId: expect.any(String),
                commandId: expect.any(String),
                siteId: expect.any(String)
            });
        });
    });

    describe('event creator', () => {
        test('should throw if given a bad command', () => {
            expect.assertions(2);
            try {
                const createLocationCommand = createTestCreateLocationCommand({ user: undefined });
                createCreateLocationEvent(createLocationCommand);
            } catch (e) {
                expect(e.name).toBe('ValidationError');
                expect(e.message).toBe('Expecting string at user but instead got: undefined.');
            }
        });

        test('should create an event if given a good commmand', () => {
            const createLocationCommand = createTestCreateLocationCommand();
            const event = createCreateLocationEvent(createLocationCommand);
            expect(event.type).toBe('LocationCreated');
        });
    });

    describe('reducer', () => {
        test('should create a location from an event', () => {
            const event = createTestCreateLocationEvent();
            const location = createLocationReducer({} as any, event);
            expect(location).toMatchSnapshot({
                created: expect.any(String)
            });
        });
    });

    describe('entry', () => {
        test('should call the entry handler', async () => {
            const command = createTestCreateLocationCommand();
            await entry({ body: JSON.stringify(command) } as any);
            expect(entryMock.requestHandler).toBeCalled();
        });
    });

    describe('integration', () => {
        test('should successfully create the location', async () => {
            const command = createTestCreateLocationCommand();
            await createLocation(command);
            const locationId = spy.mock.calls[0][0].entityId;
            const location = await locationRepo.getById(command.customerId, locationId);
            expect(location).toMatchSnapshot({
                created: expect.any(String)
            });
        });
    });
});