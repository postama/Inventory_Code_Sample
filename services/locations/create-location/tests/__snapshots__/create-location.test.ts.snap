// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create location command structure validator should pass with full structure 1`] = `
Object {
  "commandId": Any<String>,
  "customerId": Any<String>,
  "isActive": true,
  "locationId": Any<String>,
  "name": "abc",
  "siteId": Any<String>,
  "type": "CreateLocation",
  "user": "abc",
}
`;

exports[`create location integration should successfully create the location 1`] = `
Object {
  "created": Any<String>,
  "isActive": true,
  "name": "abc",
  "siteId": "TESTER",
}
`;

exports[`create location reducer should create a location from an event 1`] = `
Object {
  "created": Any<String>,
  "isActive": true,
  "name": "abc",
  "siteId": "TESTER",
}
`;
